### ISSUE TITLE
Provide training scripts

### ISSUE THREAD

**Issue by gpearlman1** [bnpy/hdp-grid-image-restoration#2]
Hi, I was going through recent denoising papers and found yours.. I see you've provided pretrained model on BSDS. Could you provide training and evaluation scripts for inpainting? Thanks.

**Comment by michaelchughes** [comment]
Thanks for your interest! What exactly do you need from an evaluation script for inpainting? Our [ICML 2017 paper (PDF)](https://www.michaelchughes.com/papers/JiHughesSudderth_ICML_2017.pdf) didn't do any quantitative evaluation, only qualitative visual comparison.

You'll see we already have code in demo.py that can perform inpainting given a pretrained model, a target image, and a specified mask. The output is a new image with the mask filled in. See here:
https://github.com/bnpy/hdp-grid-image-restoration/blob/master/demo.py#L27

**Comment by gpearlman1** [comment]
Yes, I see you're using pretrained HDP-Grid model for inpainting. I want to use HDP model trained on some other dataset and use that in inpainting task. How can I train an HDP model on new dataset?
P.S. Sorry if I'm missing something obvious. Feel free to close the issue.

**Comment by michaelchughes** [comment]
Right. You can definitely train a new model yourself on a new dataset. @gji1 can you shed more light on specific advice for training... maybe including some code?

AFAIK, the recommended high-level procedure is:

1) Extract lots of 8x8 patches from your dataset
2) Use our BNPy package [github.com/bnpy/bnpy](https://github.com/bnpy/bnpy) to fit a DP mixture model with zero-mean full covariance Gaussian likelihood (obsmodel = ZeroMeanGauss). Use birth moves / merge moves to make sure the model escapes local optima and finds good data-driven structure.
3) If you want a more powerful model, use the DP model from step 2 to initialize an HDP model and then fine-tune that fancier HDP model.
4) Apply trained model (HDP or DP) to inpainting, using code like in demo.py

**Comment by gpearlman1** [comment]
Thanks, I'll try it. Does Inpainting task take such considerable time(1.30 h in my case) for each of the testing images and do the testing images, with custom masks only have to be from BSDS? Also I see boundary pixels can't be restored. Is that the limitation of the algo or implementation?

**Comment by gji1** [comment]
Thanks for your interest, George. In our implementation the inpainting task generally takes more time to run than the denoising task because it requires more iterations. That said, you are also very welcome to tune the annealing schedule to better fit the specific data you have for better performance or faster speed. The settings we use right now are directly adopted from the public EPLL code so as to make a fair comparison with that method.

Test images don't have to be from BSDS. Actually the example image we showed in our paper is not from BSDS either. In theory, boundary pixels can also be inpainted using our algorithm. The current implementation doesn't support that yet.

**Comment by gpearlman1** [comment]
Hey,

So I'm trying to understand this paper and https://arxiv.org/abs/1209.5019
for my presentation at uni. I want to try to train the HDP inpainting model
of my own. Do I have to train a DP model first and then use it to
initialize the HDP model? Or I can just directly train HDP model?

Is it this birth and merge script of BNPy:
https://bnpy.readthedocs.io/en/latest/examples/04_bars_one_per_doc/run-05-demo=topic_model_vb+births-model=hdp_topic+mult.html#sphx-glr-examples-04-bars-one-per-doc-run-05-demo-topic-model-vb-births-model-hdp-topic-mult-py

Or this merge one:
https://bnpy.readthedocs.io/en/latest/examples/04_bars_one_per_doc/run-04-demo=topic_model_vb+merges-model=hdp_topic+mult.html#sphx-glr-examples-04-bars-one-per-doc-run-04-demo-topic-model-vb-merges-model-hdp-topic-mult-py

Or just HDP-HMM:

python -m bnpy.Run /path/to/dataset.csv HDPHMM DiagGauss memoVB --K 8

Or any other scripts as in how do I use bnpy? And also,if I want to find
PSNR and SSIM values for inpainting of this paper and EPLL, how can I do
that?

Thank you for all your help.

With Regards,
George.

**Comment by gji1** [comment]
You don't have to train a DP model first. We did that mainly because we found this would provide a better initialization. Instead, to directly start from HDP, here's the backbone script you need to build things upon:

python -m bnpy.Run /path/to/dataset HDPMixtureModel ZeroMeanGauss memoVB --K 10 --nLap 10 --nBatch 10 --moves birth,merge,shuffle

Depending on the type of your data, for best performance you may need to tune and add the optional parameters, as in the examples you mentioned. The default settings are stored in folder bnpy/bnpy/config. I suggest that you try our trained model first if your data is also just natural images.

**Comment by gpearlman1** [comment]
On which 400 images from BSDS did you train the model on? There are 200 training, 100 validation and 200 testing images available. Also what values did you choose for various hyper-parameters like concentration parameter(alpha)?
Thanks :)
